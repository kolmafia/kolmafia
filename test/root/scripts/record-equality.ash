item st = $item[seal tooth];
item sh = $item[seal-skull helmet];

print(st);
print(sh);
print("equals = " + (st == sh));
print("-------");

record rec
{
    item it;
    int count;
};

string to_string(rec r)
{
    return r.it + " (" + r.count + ")";
}

rec zero = new rec(st, 0);
rec one = new rec(st, 1);
rec two = new rec(st, 1);
rec three = new rec(sh, 1);

print("zero = " + zero);
print("one = " + one);
print("two = " + two);
print("three = " + three);
print("zero == one: " + (zero == one));
print("one == one: " + (one == one));
print("one == two: " + (one == two));
print("one == three: " + (one == three));
print("-------");

record strec {
    string str;
};

strec strec1 = new strec("ABC");
strec strec2 = new strec("abc");

print("strec1.str = " + strec1.str);
print("strec2.str = " + strec2.str);
print("strec1 == strec2: " + (strec1 == strec2));
print("strec1 â‰ˆ strec2: " + (strec1 â‰ˆ strec2));
print("-------");

typedef int[] int_array;

string to_string(int_array ia)
{
    buffer buf;
    buf.append("[");
    foreach index, val in ia {
	if (buf.length() > 1) {
	    buf.append(",");
	}
	buf.append(val);
    }
    buf.append("]");
    return buf;
}

record iarec {
    int_array ia;
};

int_array ia1 = {1, 2, 3};
iarec iarec1 = { ia: ia1 };
iarec iarec2 = { ia: ia1 };
iarec iarec3 = { ia: {1, 2, 3} };
iarec iarec4 = { ia: {1, 2, 3, 4 } };

print("iarec1.ia = " + iarec1.ia);
print("iarec2.ia = " + iarec2.ia);
print("iarec3.ia = " + iarec3.ia);
print("iarec4.ia = " + iarec4.ia);
print("iarec1 == iarec1: " + (iarec1 == iarec1));
print("iarec1 == iarec2: " + (iarec1 == iarec2));
print("iarec1 == iarec3: " + (iarec1 == iarec3));
print("iarec1 == iarec4: " + (iarec1 == iarec4));
