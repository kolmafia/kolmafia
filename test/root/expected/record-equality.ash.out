seal tooth
seal-skull helmet
equals = false
-------
zero = seal tooth (0)
one = seal tooth (1)
two = seal tooth (1)
three = seal-skull helmet (1)
zero == one: false
one == one: true
one == two: true
one == three: false
-------
strec1.str = ABC
strec2.str = abc
strec1 == strec2: false
strec1 â‰ˆ strec2: true
-------
iarec1.ia = [1,2,3]
iarec2.ia = [1,2,3]
iarec3.ia = [1,2,3]
iarec4.ia = [1,2,3,4]
iarec1 == iarec1: true
iarec1 == iarec2: true
iarec1 == iarec3: true
iarec1 == iarec4: false